std-lib/list/list.lurk

```diff
@@ -0,0 +1,62 @@
+;; a list is of the form (cons a (cons b (cons ... (cons nil)...)))
+;; some definitions for lists 
+
+;; empty_list : the empty list
+(let ((empty_list '())) (current-env))
+
+;; some functions for lists 
+
+;; list : make a list out of a (possibly unevaluated) list quote
+(letrec ((list (lambda (q)
+		 (if (car q)
+		     (cons (eval (car q)) (list (cdr q)))
+		     '()))))
+  (current-env))
```

Comment:
LIST probably requires a macro or language support to be done right (variable number of arguments).

Explicit use of EVAL is probably not desirable, but if it were, you probably want the current env to be involved.

I'm chalking this up to probably not being a useful function as written.

Reply:
I came up with that one, and i had a newer version that uses current-env, so that you can do nested lists.

```lisp
         (list (lambda (q)
                 (if (car q)
                     (cons (eval (car q) (current-env))
                           (list (cdr q)))
                     nil)))
```
I don't see any other way to produce a literal list (that's not quoted, but rather requires eval'ing args). I mean, aside from a bunch of consing. Obviously using `eval` is a terrible hack and yes `list` is supposed to be varargs.

Reply:
I think this is a clue that the concept involved is off. If we really want a function that takes a list and outputs a list, we can just use the identity function.

Normal Lisp LIST, apart from being variadic, doesn't perform explicit evaluation. It just so happens that the args are evaluated *before* LIST ever sees them.

For example, here is how SBCL defines LIST:
```
(defun list (&rest args)
  "Construct and return a list containing the objects ARGS."
  args)
```

Reply:
Yes, I understand this. 

What I really want is lisp's LIST, but I don't think that's possible to define in lurk itself right now. I'm sure you understand *why* I want it, it's for the same purpose everyone uses LIST in lisp - to make a literal list with expressions in it that I want evaluated first. Without that the only way I can think to make such a list is with repeated consing.

Reply:
Yep, I understand. I think one urge is for Lisp's LIST, which can indeed be implemented with macros or when we have support for &REST parameters.

I think you also want something that is definitionally not LIST, but that would accomplish a practical purpose. That something is basically MAP-EVAL of some flavor — which is fine.

I just would not call it LIST, since that's confusing and also will clash with some future implementation of LIST which works as it should.

Reply:
Ok yes I see your point, I started out trying to implement LIST but realized it's not possible and then kept the name on something that's not the same. What I ended up with is basically (map eval ...) like you said. I am fine with whatever it is not being in the stdlib but I will need something like this as a utility function at least until there's something better. I am ok with just copying it around. Putting it in the stdlib is probably bad because later it will have no purpose and should go away.

Reply:
I'm not even saying not have such a function. I'm just saying don't call it LIST. You could call it MAP-EVAL or something.

Reply:
Nah, if we have `map` (which we will), writing `(map eval lst)` is too trivial to be worth creating a new function.

Reply:
I think it might be a little more complicated than you want to try to inline by hand. Remember:

- EVAL isn't a function
- you need the correct environment

As a test, you want to make sure something like this works:
```
(let ((a 1))
  (do-the-thing '(a 2 (+ a 2))))
```

I think for that to work, you probably need to also pass `(current-env)` to `do-the-thing`.